File: zluqet.py
from flask import Flask, render_template, request, redirect, url_for, abort, make_response, jsonify
from flask_sqlalchemy import SQLAlchemy
from pygments import highlight
from pygments.lexers import guess_lexer, get_lexer_by_name
from pygments.formatters import HtmlFormatter
from pygments.util import ClassNotFound
import string
import random
from datetime import datetime
import time
from collections import defaultdict, deque
from sqlalchemy.engine import Engine
from sqlalchemy import event
from types import SimpleNamespace 
from cachetools import LRUCache
from werkzeug.exceptions import RequestEntityTooLarge

app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///pastes.db'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
app.config['SQLALCHEMY_EXPIRE_ON_COMMIT'] = False
app.config['MAX_CONTENT_LENGTH'] = 1024 * 1024  

app.config['SQLALCHEMY_ENGINE_OPTIONS'] = {
    'pool_size': 10,
    'max_overflow': 20,
    'pool_pre_ping': True,
}
app.secret_key = 'zluqet'

db = SQLAlchemy(app)

@event.listens_for(Engine, "connect")
def set_sqlite_pragma(dbapi_connection, connection_record):
    cursor = dbapi_connection.cursor()
    cursor.execute("PRAGMA synchronous = OFF")
    cursor.execute("PRAGMA journal_mode = WAL")
    cursor.execute("PRAGMA cache_size = 10000")
    cursor.close()

ip_requests = defaultdict(deque)
EXEMPT_IPS = {}
REQUESTS_PER_MINUTE = 5

@app.before_request
def rate_limit():
    if request.path.startswith('/api/'):
        client_ip = request.remote_addr
        if client_ip in EXEMPT_IPS:
            return
        
        now = time.time()
        timestamps = ip_requests[client_ip]
        while timestamps and now - timestamps[0] > 60:
            timestamps.popleft()

        if len(timestamps) >= REQUESTS_PER_MINUTE:
            return jsonify({'error': 'Too many requests, please slow down.'}), 429
        
        timestamps.append(now)

class Paste(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    content = db.Column(db.Text, nullable=False)
    key = db.Column(db.String(8), unique=True, nullable=False)
    created_at = db.Column(db.DateTime, nullable=False, default=datetime.utcnow)

    def __repr__(self):
        return f'<Paste {self.key}>'

# Use LRU caches to limit memory usage
paste_cache = LRUCache(maxsize=1000) 
highlight_cache = LRUCache(maxsize=1000) 

def cache_paste(p):
    return SimpleNamespace(
        content=p.content,
        key=p.key,
        created_at=p.created_at
    )

def get_paste_by_key(key):
    if key in paste_cache:
        return paste_cache[key]
    paste = Paste.query.filter_by(key=key).first()
    if paste:
        lite = cache_paste(paste)
        paste_cache[key] = lite
        return lite
    return None

formatter = HtmlFormatter(style="default")
STYLE_DEFS = formatter.get_style_defs('.highlight')

def generate_key(length=8):
    chars = string.ascii_uppercase + string.digits
    while True:
        key = ''.join(random.choices(chars, k=length))
        if not Paste.query.filter_by(key=key).first():
            return key

def precompute_highlighting(content, key):
    try:
        lexer = guess_lexer(content)
    except ClassNotFound:
        lexer = get_lexer_by_name("text", stripall=True)
    highlighted = highlight(content, lexer, formatter)
    highlight_cache[key] = highlighted

# Maximum allowed characters for a paste
MAX_LENGTH = 100000

@app.route('/', methods=['GET', 'POST'])
def index():
    if request.method == 'POST':
        content = request.form.get('content')
        if not content:
            return redirect(url_for('index'))
        
        if len(content) > MAX_LENGTH:
            error_message = (
                f"Your paste exceeds the maximum allowed character limit "
                f"of {MAX_LENGTH} characters. Please reduce your content before saving."
            )
            # Re-render the page with the error message and user's content preserved.
            return render_template('index.html', error=error_message, content=content)
        
        key = generate_key()
        new_paste = Paste(content=content, key=key)
        db.session.add(new_paste)
        db.session.commit()
        
        lite = cache_paste(new_paste)
        paste_cache[key] = lite
        precompute_highlighting(new_paste.content, key)
        
        return redirect(url_for('view_paste', key=key))
    
    return render_template('index.html')

@app.route('/<key>')
def view_paste(key):
    paste = get_paste_by_key(key)
    if not paste:
        abort(404)
    highlighted_code = highlight_cache.get(key)
    if not highlighted_code:
        try:
            lexer = guess_lexer(paste.content)
        except ClassNotFound:
            lexer = get_lexer_by_name("text", stripall=True)
        highlighted_code = highlight(paste.content, lexer, formatter)
        highlight_cache[key] = highlighted_code

    response = make_response(
        render_template(
            'view_paste.html',
            paste=paste,
            highlighted_code=highlighted_code,
            style_defs=STYLE_DEFS
        )
    )
    response.headers['X-Robots-Tag'] = 'noindex, nofollow'
    return response

@app.route('/raw/<key>')
def raw_paste(key):
    paste = get_paste_by_key(key)
    if not paste:
        abort(404)
    return paste.content, {'Content-Type': 'text/plain'}

@app.route('/edit/<key>', methods=['GET', 'POST'])
def dupe_paste(key):
    paste = get_paste_by_key(key)
    if not paste:
        abort(404)

    if request.method == 'POST':
        content = request.form.get('content')
        if not content:
            return redirect(url_for('index'))

        if len(content) > MAX_LENGTH:
            error_message = (
                f"Your paste exceeds the maximum allowed character limit "
                f"of {MAX_LENGTH} characters. Please reduce your content before saving."
            )
            return render_template('edit_paste.html', error=error_message, paste=paste, content=content)

        new_key = generate_key()
        new_paste = Paste(content=content, key=new_key)
        db.session.add(new_paste)
        db.session.commit()
        
        lite = cache_paste(new_paste)
        paste_cache[new_key] = lite
        precompute_highlighting(new_paste.content, new_key)
        
        return redirect(url_for('view_paste', key=new_key))

    return render_template('edit_paste.html', paste=paste)

@app.route('/api/documents', methods=['POST'])
def api_create_paste():
    text_content = request.get_data(as_text=True)
    if not text_content:
        return jsonify({'error': 'No content provided.'}), 400

    if len(text_content) > MAX_LENGTH:
        return jsonify({'error': f"Your paste exceeds the maximum allowed character limit of {MAX_LENGTH} characters."}), 400

    key = generate_key()
    new_paste = Paste(content=text_content, key=key)
    db.session.add(new_paste)
    db.session.commit()
    
    lite = cache_paste(new_paste)
    paste_cache[key] = lite
    precompute_highlighting(new_paste.content, key)

    response_data = {'key': key}
    return jsonify(response_data), 200

@app.route('/api/documents/<key>', methods=['GET'])
def api_get_paste(key):
    paste = get_paste_by_key(key)
    if not paste:
        return jsonify({'error': 'Paste not found.'}), 404
    return jsonify({
        'key': paste.key,
        'content': paste.content,
        'created_at': paste.created_at.isoformat()
    })

@app.errorhandler(404)
def page_not_found(e):
    return render_template('errors/404.html'), 404

@app.errorhandler(RequestEntityTooLarge)
def handle_request_entity_too_large(error):
    error_message = (
        f"Your paste is too large! The maximum allowed size is {MAX_LENGTH} characters. "
        "Please reduce your content and try again."
    )
    return render_template("errors/too_large.html", error=error_message), 413

if __name__ == '__main__':
    with app.app_context():
        db.create_all()
    app.run(debug=False, threaded=True)

File: zluqet.js
// zluqet.js

const express      = require('express');
const path         = require('path');
const Sequelize    = require('sequelize');
const { DataTypes }= Sequelize;
const bodyParser   = require('body-parser');
const cookieParser = require('cookie-parser');
const { LRUCache } = require('lru-cache');

const app = express();
app.set('trust proxy', 1);

// —————————————
// Static files
// —————————————
app.use(express.static(path.join(__dirname, 'static')));

// —————————————
// View engine
// —————————————
app.set('views', path.join(__dirname, 'templates'));
app.set('view engine', 'ejs');

// —————————————
// Middleware
// —————————————
app.use(cookieParser());
app.use(bodyParser.urlencoded({ extended: false, limit: '1mb' }));
app.use(bodyParser.text({ type: 'text/plain', limit: '1mb' }));
app.use(bodyParser.json({ limit: '1mb' }));

// —————————————
// Database (SQLite + Sequelize)
// —————————————
const sequelize = new Sequelize({
  dialect: 'sqlite',
  storage: path.join(__dirname, 'instance/pastes.db'),
  logging: false,
  pool: { max: 10, min: 0, acquire: 30000, idle: 10000 }
});

sequelize.authenticate()
  .then(async () => {
    await sequelize.query('PRAGMA synchronous = OFF');
    await sequelize.query('PRAGMA journal_mode = WAL');
    await sequelize.query('PRAGMA cache_size = 10000');
  })
  .catch(console.error);

const Paste = sequelize.define('Paste', {
  id:         { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
  content:    { type: DataTypes.TEXT,    allowNull: false },
  key:        { type: DataTypes.STRING(8), unique: true, allowNull: false },
  created_at: { type: DataTypes.DATE,    allowNull: false, defaultValue: Sequelize.NOW }
}, {
  tableName: 'pastes',
  timestamps: false
});

sequelize.sync();

// —————————————
// Rate limiting (simple IP-based for /api/)
// —————————————
const ipRequests = new Map();
const EXEMPT_IPS = new Set();
const REQUESTS_PER_MINUTE = 5;

app.use((req, res, next) => {
  if (req.path.startsWith('/api/')) {
    const ip = req.ip;
    if (!EXEMPT_IPS.has(ip)) {
      const now = Date.now() / 1000;
      if (!ipRequests.has(ip)) ipRequests.set(ip, []);
      const timestamps = ipRequests.get(ip);
      while (timestamps.length && now - timestamps[0] > 60) {
        timestamps.shift();
      }
      if (timestamps.length >= REQUESTS_PER_MINUTE) {
        return res.status(429).json({ error: 'Too many requests, please slow down.' });
      }
      timestamps.push(now);
    }
  }
  next();
});

// —————————————
// In-memory caching for database lookups
// —————————————
const pasteCache = new LRUCache({ max: 1000 });

function cachePaste(p) {
  return { content: p.content, key: p.key, created_at: p.created_at };
}

async function getPasteByKey(key) {
  if (pasteCache.has(key)) {
    return pasteCache.get(key);
  }
  const p = await Paste.findOne({ where: { key } });
  if (!p) return null;
  const lite = cachePaste(p);
  pasteCache.set(key, lite);
  return lite;
}

// —————————————
// Utility: generate a unique 8-char key
// —————————————
async function generateKey(length = 8) {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  let key, exists;
  do {
    key = Array.from({ length },
      () => chars[Math.floor(Math.random() * chars.length)]
    ).join('');
    exists = await Paste.findOne({ where: { key } });
  } while (exists);
  return key;
}

const MAX_LENGTH = 25000;

// —————————————
// Routes
// —————————————

// Home / New Paste
app.get('/', (req, res) => {
  res.render('index', { content: '', error: null });
});

app.post('/', async (req, res) => {
  const content = req.body.content;
  if (!content) return res.redirect('/');
  if (content.length > MAX_LENGTH) {
    return res.status(400).render('index', {
      content,
      error: `Your paste exceeds the maximum allowed character limit of ${MAX_LENGTH} characters. Please reduce your content before saving.`
    });
  }

  const key = await generateKey();
  const newPaste = await Paste.create({ content, key });
  pasteCache.set(key, cachePaste(newPaste));
  res.redirect(`/${key}`);
});

// View Paste (HTML)
app.get('/:key', async (req, res, next) => {
  const paste = await getPasteByKey(req.params.key);
  if (!paste) return next();

  // Format created_at for display
  const dt = new Date(paste.created_at);
  const opts = {
    month:   'short',
    day:     '2-digit',
    year:    'numeric',
    hour:    '2-digit',
    minute:  '2-digit',
    hour12:  false
  };
  const formattedDate = dt.toLocaleString('en-US', opts).replace(',', '');

  res.set('X-Robots-Tag', 'noindex, nofollow');
  res.render('view_paste', { paste, formattedDate });
});

// View Paste (raw text)
app.get('/raw/:key', async (req, res, next) => {
  const paste = await getPasteByKey(req.params.key);
  if (!paste) return next();
  res.type('text/plain').send(paste.content);
});

// Edit / Duplicate Paste
app.get('/edit/:key', async (req, res, next) => {
  const paste = await getPasteByKey(req.params.key);
  if (!paste) return next();
  res.render('edit_paste', { paste, error: null });
});

app.post('/edit/:key', async (req, res, next) => {
  const original = await getPasteByKey(req.params.key);
  if (!original) return next();

  const content = req.body.content;
  if (!content) return res.redirect('/');
  if (content.length > MAX_LENGTH) {
    return res.status(400).render('edit_paste', {
      paste: original,
      error: `Your paste exceeds the maximum allowed character limit of ${MAX_LENGTH} characters. Please reduce your content before saving.`
    });
  }

  const newKey = await generateKey();
  const newPaste = await Paste.create({ content, key: newKey });
  pasteCache.set(newKey, cachePaste(newPaste));
  res.redirect(`/${newKey}`);
});

// API: create a paste
app.post('/api/documents', async (req, res) => {
  const text_content = req.body;
  if (!text_content) {
    return res.status(400).json({ error: 'No content provided.' });
  }
  if (text_content.length > MAX_LENGTH) {
    return res.status(400).json({ error: `Your paste exceeds the maximum allowed character limit of ${MAX_LENGTH} characters.` });
  }

  const key = await generateKey();
  const newPaste = await Paste.create({ content: text_content, key });
  pasteCache.set(key, cachePaste(newPaste));
  res.json({ key });
});

// API: retrieve a paste
app.get('/api/documents/:key', async (req, res) => {
  const paste = await getPasteByKey(req.params.key);
  if (!paste) {
    return res.status(404).json({ error: 'Paste not found.' });
  }
  res.json({
    key:        paste.key,
    content:    paste.content,
    created_at: paste.created_at.toISOString()
  });
});

// 404 handler
app.use((req, res) => {
  res.status(404).render('errors/404');
});

// Error handler for oversized bodies
app.use((err, req, res, next) => {
  if (err.status === 413 || err.type === 'entity.too.large') {
    return res.status(413).render('errors/too_large', {
      error: `Your paste is too large! The maximum allowed size is ${MAX_LENGTH} characters. Please reduce your content and try again.`
    });
  }
  next(err);
});

// Start the server
const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server listening on port ${PORT}`);
});
